File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/literal/NumberLiteral.java
package me.itzisonn_.meazy.parser.ast.expression.literal;

import lombok.Getter;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

import java.lang.constant.ConstantDescs;

@Getter
@NullMarked
public class NumberLiteral implements Expression {
    private final String value;

    public NumberLiteral(String value) {
        this.value = value;
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        try {
            int number = Integer.parseInt(value);
            instructionsSet.loadConstant(number);
        }
        catch (NumberFormatException e) {
            double number = Double.parseDouble(value);
            instructionsSet.loadConstant(number);
        }
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        try {
            Integer.parseInt(value);
            return DataType.ofNonNull(ConstantDescs.CD_int);
        }
        catch (NumberFormatException e) {
            return DataType.ofNonNull(ConstantDescs.CD_double);
        }
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/expression/literal/NumberLiteral.kt
package me.itzisonn_.meazy.parser.ast.expression.literal

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.runtime.data.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.ast.SymbolMap
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.runtime.environment.Environment
import java.lang.constant.ConstantDescs

class NumberLiteral(val value: String) : Expression {
    override val children = setOf<ProgramUnit>()

    context(parents: ParentMap, symbols: SymbolMap)
    override fun emit(instructions: InstructionsSet, environment: Environment) {
        val int = value.toIntOrNull()
        if (int != null) {
            instructions.loadConstant(int)
            return
        }

        val double = value.toDoubleOrNull() ?: error("Invalid number $value")
        instructions.loadConstant(double)
    }

    context(parents: ParentMap)
    override fun getType(environment: Environment): DataType {
        val classDesc = when {
            value.toIntOrNull() != null -> ConstantDescs.CD_int
            value.toDoubleOrNull() != null -> ConstantDescs.CD_double
            else -> error("Invalid number $value")
        }

        return DataType.ofNonNull(classDesc)
    }
}
